src/MagnumPlugins/BasisImageConverter/Test/BasisImageConverterTest.cpp

```diff
@@ -0,0 +1,177 @@
+        _converterManager.instantiate("BasisImageConverter");
+
+    std::ostringstream out;
+    Error redirectError{&out};
+
+    converter->configuration().setValue("num_threads", 0);
+    CORRADE_VERIFY(!converter->exportToData(ImageView2D{PixelFormat::RGB8Unorm, {}, nullptr}));
+    CORRADE_COMPARE(out.str(),
+        "Trade::BasisImageConverter::exportToData(): invalid configuration: num_threads is expected to be at least 1, but was: 0\n");
+}
+
+constexpr const char OriginalRgbData[] = {
+    /* Skip */
+    0, 0, 0, 0, 0, 0, 0, 0,
+
+    1, 2, 3, 2, 3, 4, 0, 0,
+    3, 4, 5, 4, 5, 6, 0, 0,
+    5, 6, 7, 6, 7, 8, 0, 0
+};
+
+const ImageView2D OriginalRgb{PixelStorage{}.setSkip({0, 1, 0}),
+    PixelFormat::RGB8Unorm, {2, 3}, OriginalRgbData};
+
+void BasisImageConverterTest::rgb() {
+    const auto data = _converterManager.instantiate("BasisImageConverter")->exportToData(OriginalRgb);
+    CORRADE_VERIFY(data);
+
+    if(_importerManager.loadState("BasisImporter") == PluginManager::LoadState::NotFound)
+        CORRADE_SKIP("BasisImporter plugin not found, cannot test");
+
+    Containers::Pointer<AbstractImporter> importer = _importerManager.instantiate("BasisImporter");
+    importer->configuration().setValue("format", "Etc2");
+    CORRADE_VERIFY(importer->openData(data));
+    Containers::Optional<Trade::ImageData2D> converted = importer->image2D(0);
+    CORRADE_VERIFY(converted);
+
+    CORRADE_COMPARE(converted->size(), Vector2i(2, 3));
+    /* Ensure there is no alpha channel
+       In case pixel data is not copied properly to basis, not all alpha values will be
+       opaque and basis will decide that the images has an alpha channel on export */
+    CORRADE_COMPARE(converted->compressedFormat(), CompressedPixelFormat::Etc2RGB8Unorm);
+}
+
+constexpr const char OriginalRgbaData[] = {
+    /* Skip */
+    0, 0, 0, 0, 0, 0, 0, 0,
+
+    1, 2, 3, 2, 3, 4, 0, 0,
+    3, 4, 5, 4, 5, 6, 0, 0,
+    5, 6, 7, 6, 7, 8, 0, 0,
+    1, 2, 3, 4, 5, 6, 7, 8
+};
+
+const ImageView2D OriginalRgba{PixelStorage{}.setSkip({0, 1, 0}),
+    PixelFormat::RGBA8Unorm, {2, 3}, OriginalRgbaData};
+
+void BasisImageConverterTest::rgba() {
+    const auto data = _converterManager.instantiate("BasisImageConverter")->exportToData(OriginalRgba);
+    CORRADE_VERIFY(data);
+
+    if(_importerManager.loadState("BasisImporter") == PluginManager::LoadState::NotFound)
+        CORRADE_SKIP("BasisImporter plugin not found, cannot test");
+
+    Containers::Pointer<AbstractImporter> importer = _importerManager.instantiate("BasisImporter");
+    importer->configuration().setValue("format", "Etc2");
+    CORRADE_VERIFY(importer->openData(data));
+    Containers::Optional<Trade::ImageData2D> converted = importer->image2D(0);
+    CORRADE_VERIFY(converted);
+
+    CORRADE_COMPARE(converted->size(), Vector2i(2, 3));
+    /* Ensure there is an alpha channel */
+    CORRADE_COMPARE(converted->compressedFormat(), CompressedPixelFormat::Etc2RGBA8Unorm);
+}
+
```

Comment:
Is there any chance to trigger the encoder failures? Otherwise the error code enum is never covered.

Reply:
`cECFailedReadingSourceImages` is output on dimension issues (zero, which we catch before and larger than max dimensions, which would cause a pretty large allocation during test?), too many output slices, mipmap generation failure and during final sanity checks. I didn't find a obvious way to provoke any of those, though

`cECFailedValidating` could be provoked by settings texture type to cube map and passing non divisible by 6 amount of images.

`cECFailedFrontEnd` can be triggered with `m_params.m_max_endpoint_clusters >`basisu_frontend::cMaxEndpointClusters`, in case that isn't also checked beforehand.

`cECFailedBackend` has loads of code paths that lead to an error, but this code is accordingly complex and most code paths seem more like sanity checks than actual errors, maybe through the selector codebook parameter or something, but I think most of that is checked beforehand.

`cECFailedCreateBasisFile` checksum or transcoding failures, not sure how we'd produce one of these without messing with Basis' memory while it's running or creating a custom alloc and feeding that in or something 🤔

Reply:
For the first we could create an impossibly large ImageView with sized-but-null data view. That could work I think?

Reply:
That crashes before it even reaches .process() :D Especially since the data needs to be copied anyway

Reply:
`cECFailedValidating` => we don't expose texture type because we can't input more than one image anyway. I don't think there is a way to provoke this one either.

`cECFailedFrontEnd` => ✅ 

And I think that's the only one.

Reply:
ookay, skip that, then :)